File: fe/mobile/android/app/src/main/java/com/capstone/ui/viewmodel/NotificationViewModel.java
package com.capstone.ui.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.capstone.domain.model.Notification;
import com.capstone.domain.repository.NotificationRepository;
import dagger.hilt.android.lifecycle.HiltViewModel;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import javax.inject.Inject;

@HiltViewModel
public class NotificationViewModel extends ViewModel {
    private final NotificationRepository notificationRepository;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    private final MutableLiveData<List<Notification>> notifications = new MutableLiveData<>(new ArrayList<>());
    private final MutableLiveData<Boolean> isLoading = new MutableLiveData<>(false);
    private final MutableLiveData<String> error = new MutableLiveData<>(null);

    @Inject
    public NotificationViewModel(NotificationRepository notificationRepository) {
        this.notificationRepository = notificationRepository;
    }

    public LiveData<List<Notification>> getNotifications() { return notifications; }
    public LiveData<Boolean> getIsLoading() { return isLoading; }
    public LiveData<String> getError() { return error; }

    public void fetchNotifications(int page, int size) {
        isLoading.setValue(true);
        executor.execute(() -> {
            try {
                List<Notification> newNotifications = notificationRepository.getNotifications(page, size);
                List<Notification> currentList = new ArrayList<>(notifications.getValue());
                if (page == 0) {
                    notifications.postValue(newNotifications);
                } else {
                    currentList.addAll(newNotifications);
                    notifications.postValue(currentList);
                }
                error.postValue(null);
            } catch (Exception e) {
                error.postValue(e.getMessage());
            } finally {
                isLoading.postValue(false);
            }
        });
    }

    public void markAsRead(String id) {
        executor.execute(() -> {
            try {
                boolean success = notificationRepository.markAsRead(id);
                if (success) {
                    List<Notification> currentList = new ArrayList<>(notifications.getValue());
                    for (int i = 0; i < currentList.size(); i++) {
                        Notification n = currentList.get(i);
                        if (n.getId().equals(id)) {
                            currentList.set(i, n.copyWithRead(true));
                            break;
                        }
                    }
                    notifications.postValue(currentList);
                }
            } catch (Exception e) {
                error.postValue(e.getMessage());
            }
        });
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        executor.shutdown();
    }
}

File: fe/mobile/android/app/src/main/java/com/capstone/ui/viewmodel/NotificationViewModel.kt
package com.capstone.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.capstone.domain.model.Notification
import com.capstone.domain.repository.NotificationRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class NotificationViewModel @Inject constructor(
    private val notificationRepository: NotificationRepository
) : ViewModel() {

    private val _notifications = MutableStateFlow<List<Notification>>(emptyList())
    val notifications: StateFlow<List<Notification>> = _notifications

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error

    /**
     * Lấy danh sách thông báo có phân trang
     */
    fun fetchNotifications(page: Int, size: Int = 20) {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                val newNotifications = notificationRepository.getNotifications(page, size)
                _notifications.value = if (page == 0) newNotifications else _notifications.value + newNotifications
                _error.value = null
            } catch (e: Exception) {
                _error.value = e.message
            } finally {
                _isLoading.value = false
            }
        }
    }

    /**
     * Đánh dấu đã đọc
     */
    fun markAsRead(id: String) {
        viewModelScope.launch {
            try {
                val success = notificationRepository.markAsRead(id)
                if (success) {
                    // Update local state
                    _notifications.value = _notifications.value.map {
                        if (it.id == id) it.copy(isRead = true) else it
                    }
                }
            } catch (e: Exception) {
                _error.value = e.message
            }
        }
    }
}
